pkg/controller/storagecluster/reconcile.go

```diff
@@ -671,11 +674,14 @@ func determineFailureDomain(topologyMap *ocsv1.NodeTopologyMap) string {
 		if strings.Contains(label, "zone") {
 			if len(labelValues) >= 3 {
 				failureDomain = "zone"
-				break
 			}
 		}
+		// If we have already assigned a rack-based topology to the cluster,
+		// this must be maintained.
+		if strings.Contains(label, "rack") {
+			return "rack"
+		}
```

Comment:
@egafford, somehow, when discussing this change, I had imagined that we had somehow recorded the failure domain setting in the resulting storage cluster and could somehow retrieve that. But I see what your approach is: I think if I understand your change correctly, you are saying if we see at least one node with a "rack" label, then we have started with rack once, and we are sticking with rack. I see how that should work.

 But it's also has potential problems, introduces a slight risk for bugs for future changes: We currently decide to put rack labels only if we don't do zone. But what if later, we decide to always put rack lables? Or what if we want to define a way to move from zone to rack (and back again?) So the current approach is simple enough, but we need to keep in mind for future changes that this could would need to change if we ever decide to always add rack labels etc. 

Given the simplicity of the patch, I agree that we should probably not do any more complicated approach right now. Possibly note in a comment that it extrapolates from possibly still incomplete label data (`ensureNodeRacks` is invoked later), and assumes knowledge of how the other code works. But maybe I'm just overly paranoid... :wink:

Reply:
No, I see your point. Happily, I think we can absolutely introduce a path to always put rack labels on from here, and racks are inherently zoned, so I don't think we're boxing ourselves in. If in a future release we decide that each zone always consists of racks of n machines, that's easy enough to implement if we're already using zone as the failure domain.

I also agree that a fuller refactor of this code for clarity might help matters (as there absolutely are some assumptions built into my approach to preserve method signatures), but I think that's definitely a between-releases task. For release stabilization, I try to go minimalist. :)